gk/fib.c

```diff
@@ -0,0 +1,1343 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "gatekeeper_fib.h"
+#include "gatekeeper_gk.h"
+#include "gatekeeper_lls.h"
+#include "gatekeeper_main.h"
+
+void
+destroy_neigh_hash_table(struct neighbor_hash_table *neigh)
+{
+	if (neigh->cache_tbl != NULL) {
+		rte_free(neigh->cache_tbl);
+		neigh->cache_tbl = NULL;
+	}
+
+	if (neigh->hash_table != NULL) {
+		rte_hash_free(neigh->hash_table);
+		neigh->hash_table = NULL;
+	}
+}
+
+static int
+gk_lpm_add_ipv4_routes(struct ipv4_lpm_route *routes,
+	unsigned int num_routes, struct gk_lpm *ltbl)
+{
+	int ret = 0;
+	unsigned int i;
+	for (i = 0; i < num_routes; i++) {
+		ret = rte_lpm_add(ltbl->lpm, ntohl(routes[i].ip),
+			routes[i].depth, routes[i].nexthop);
+		if (ret < 0)
+			goto out;
+		ltbl->fib_tbl[routes[i].nexthop].ref_cnt++;
+	}
+
+out:
+	return ret;
+}
+
+static int
+gk_lpm_add_ipv6_routes(struct ipv6_lpm_route *routes,
+	unsigned int num_routes, struct gk_lpm *ltbl)
+{
+	int ret = 0;
+	unsigned int i;
+	for (i = 0; i < num_routes; i++) {
+		ret = rte_lpm6_add(ltbl->lpm6, routes[i].ip,
+			routes[i].depth, routes[i].nexthop);
+		if (ret < 0)
+			goto out;
+		ltbl->fib_tbl6[routes[i].nexthop].ref_cnt++;
+	}
+
+out:
+	return ret;
+}
+
+static int
+parse_ip_prefix(const char *ip_prefix, struct ipaddr *res)
+{
+	/* Need to make copy to tokenize. */
+	size_t ip_prefix_len = strlen(ip_prefix);
+	char ip_prefix_copy[ip_prefix_len + 1];
+	char *ip_addr;
+
+	char *saveptr;
+	char *prefix_len_str;
+	char *end;
+	long prefix_len;
+	int ip_type;
+
+	strncpy(ip_prefix_copy, ip_prefix, ip_prefix_len + 1);
+
+	ip_addr = strtok_r(ip_prefix_copy, "/", &saveptr);
+	if (ip_addr == NULL)
+		return -1;
+
+	ip_type = get_ip_type(ip_addr);
+	if (ip_type != AF_INET && ip_type != AF_INET6)
+		return -1;
+
+	prefix_len_str = strtok_r(NULL, "\0", &saveptr);
+	if (prefix_len_str == NULL)
+		return -1;
+
+	prefix_len = strtol(prefix_len_str, &end, 10);
+	if (prefix_len_str == end || !*prefix_len_str || *end) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: prefix length \"%s\" is not a number\n",
+			prefix_len_str);
+		return -1;
+	}
+
+	if ((prefix_len == LONG_MAX || prefix_len == LONG_MIN) &&
+			errno == ERANGE) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: prefix length \"%s\" caused underflow or overflow\n",
+			prefix_len_str);
+		return -1;
+	}
+
+	if (prefix_len < 0 || prefix_len > max_prefix_len(ip_type)) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: prefix length \"%s\" is out of range\n",
+			prefix_len_str);
+		return -1;
+	}
+
+	if (convert_str_to_ip(ip_addr, res) < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to convert ip prefix %s to a number!\n",
+			ip_prefix);
+		return -1;
+	}
+
```

Comment:
I would rather to choose the option 1, since if we rewrite ``parse_ip_prefix()``, the function ``max_prefix_len()`` still needs the ``ip_type`` information, which is different from ``res->proto``. Moreover, if we change ``max_prefix_len()`` to meet this need, we also need to change other piece of code in ``lib/net.c``.